s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/input/AvroTransformer.java

```diff
@@ -66,18 +66,67 @@ public byte[] getValueBytes(final Object record, final String topic, final S3Sou
 
     private Stream<Object> readAvroRecordsAsStream(final IOSupplier<InputStream> inputStreamIOSupplier,
             final DatumReader<GenericRecord> datumReader) {
-        try (DataFileStream<GenericRecord> dataFileStream = new DataFileStream<>(inputStreamIOSupplier.get(),
-                datumReader)) {
-            // Wrap DataFileStream in a Stream using a Spliterator for lazy processing
-            return StreamSupport.stream(
-                    Spliterators.spliteratorUnknownSize(dataFileStream, Spliterator.ORDERED | Spliterator.NONNULL),
-                    false);
+        InputStream inputStream; // NOPMD CloseResource: being closed in try resources iterator
+        DataFileStream<GenericRecord> dataFileStream; // NOPMD CloseResource: being closed in try resources iterator
+        try {
+            // Open input stream from S3
+            inputStream = inputStreamIOSupplier.get();
+
+            // Ensure the DataFileStream is initialized correctly with the open stream
+            dataFileStream = new DataFileStream<>(inputStream, datumReader);
+
+            // Wrap DataFileStream in a Stream using a custom Spliterator for lazy processing
+            return StreamSupport.stream(new AvroRecordSpliterator<>(dataFileStream), false).onClose(() -> {
+                try {
+                    inputStream.close(); // Ensure the reader is closed after streaming
```

Comment:
dataFileStream.close() will close the inputstream as well, but if anything changes it will make sure anything in the dataFileStream is handled as well.

Reply:
Thanks, updated.